src/Map.ts

```diff
@@ -0,0 +1,825 @@
+}
+
+/**
+ * Test whether or not one Map contains all of the keys and values contained in another Map
+ *
+ * @since 1.14.0
+ */
+export const isSubmap = <K, A>(SK: Setoid<K>, SA: Setoid<A>): ((d1: Map<K, A>, d2: Map<K, A>) => boolean) => {
+  const lookupWithKeyS = lookupWithKey(SK)
+  return (d1: Map<K, A>, d2: Map<K, A>): boolean => {
+    const entries = d1.entries()
+    let e: IteratorResult<[K, A]>
+    while (!(e = entries.next()).done) {
+      const [k, a] = e.value
+      const d1OptA = lookupWithKeyS(k, d1)
+      const d2OptA = lookupWithKeyS(k, d2)
+      if (d1OptA.isNone() || d2OptA.isNone() || !SK.equals(k, d2OptA.value[0]) || !SA.equals(a, d2OptA.value[1])) {
+        return false
+      }
+    }
+    return true
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+export const empty = new Map<never, never>()
+
+/**
+ * @since 1.14.0
+ */
+export const getSetoid = <K, A>(SK: Setoid<K>, SA: Setoid<A>): Setoid<Map<K, A>> => {
+  const isSubmap_ = isSubmap(SK, SA)
+  return fromEquals((x, y) => isSubmap_(x, y) && isSubmap_(y, x))
+}
+
+/**
+ * Gets {@link Monoid} instance for Maps given {@link Semigroup} instance for their values
+ *
+ * @since 1.14.0
+ */
+export const getMonoid = <K, A>(SK: Setoid<K>, SA: Semigroup<A>): Monoid<Map<K, A>> => {
+  return {
+    concat: (mx, my) => {
+      const r = new Map(mx)
+      const entries = my.entries()
+      let e: IteratorResult<[K, A]>
+      while (!(e = entries.next()).done) {
+        const [k, a] = e.value
+        const lookupWithKeyS = lookupWithKey(SK)
+        const mxOptA = lookupWithKeyS(k, mx)
+        const myOptA = lookupWithKeyS(k, my)
+        r.set(
+          k,
+          mxOptA.isSome() && myOptA.isSome() && SK.equals(k, myOptA.value[0]) ? SA.concat(mxOptA.value[1], a) : a
+        )
+      }
+      return r
+    },
+    empty
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+function filter<K, A, B extends A>(fa: Map<K, A>, p: Refinement<A, B>): Map<K, B>
+function filter<K, A>(fa: Map<K, A>, p: Predicate<A>): Map<K, A>
+function filter<K, A>(fa: Map<K, A>, p: Predicate<A>): Map<K, A> {
+  return filterWithKey(fa, (_, a) => p(a))
+}
+
+/**
+ * @since 1.14.0
+ */
+const mapWithKey = <K, A, B>(fa: Map<K, A>, f: (k: K, a: A) => B): Map<K, B> => {
+  const m = new Map<K, B>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [key, a] = e.value
+    m.set(key, f(key, a))
+  }
+  return m
+}
+
+/**
+ * @since 1.14.0
+ */
+const map_ = <K, A, B>(fa: Map<K, A>, f: (a: A) => B): Map<K, B> => mapWithKey(fa, (_, a) => f(a))
+
+/**
+ * @since 1.14.0
+ */
+const reduce = <K>(O: Ord<K>): (<A, B>(fa: Map<K, A>, b: B, f: (b: B, a: A) => B) => B) => {
+  return (fa, b, f) => reduceWithKey(O)(fa, b, (_, b, a) => f(b, a))
+}
+
+/**
+ * @since 1.14.0
+ */
+const foldMap = <K>(O: Ord<K>): (<M>(M: Monoid<M>) => <A>(fa: Map<K, A>, f: (a: A) => M) => M) => M => (fa, f) =>
+  foldMapWithKey(O)(M)(fa, (_, a) => f(a))
```

Comment:
`foldMapWithKey(O)` can be moved outside

Reply:
I made this change. Marking it as resolved.